packages/markers/src/browser/problem/problem-decorator.ts

```diff
@@ -168,3 +171,131 @@ export namespace ProblemDecorator {
         (left.data.severity || Number.MAX_SAFE_INTEGER) - (right.data.severity || Number.MAX_SAFE_INTEGER);
 
 }
+
+@injectable()
+export class ProblemTabBarDecorator implements TabBarDecorator {
+
+    readonly id = 'theia-problem-tabbar-decorator';
+
+    protected emitter: Emitter<(titles: Title<Widget>[]) => void>;
+
+    @inject(ProblemManager)
+    protected readonly problemManager: ProblemManager;
+    @inject(WorkspaceService)
+    protected readonly workspaceService: WorkspaceService;
+
+    @postConstruct()
+    protected init(): void {
+        this.emitter = new Emitter();
+        this.problemManager.onDidChangeMarkers(() => this.fireDidChangeDecorations((titles: Title<Widget>[]) => this.collectDecorators(titles)));
+    }
+
+    decorate(titles: Title<Widget>[]): Map<string, TreeDecoration.Data> {
+        return this.collectDecorators(titles);
+    }
+
+    get onDidChangeDecorations(): Event<(titles: Title<Widget>[]) => void> {
+        return this.emitter.event;
+    }
+
+    protected fireDidChangeDecorations(event: (titles: Title<Widget>[]) => Map<string, TabBarDecoration.Data>): void {
+        this.emitter.fire(event);
+    }
+
+    /**
+     * Collect decorators for the tabs.
+     * @returns {Map<string, TabBarDecoration.Data>} A map from the tab URI to the tab decoration data.
+     */
+    protected collectDecorators(titles: Title<Widget>[]): Map<string, TabBarDecoration.Data> {
+        const result: Map<string, Marker<Diagnostic>> = new Map();
+        const markers = this.groupMarkersByURI(this.collectMarkers());
+        for (const title of titles) {
+            const relativeUri = title.caption;
+            const root = this.workspaceService.tryGetRoots()[0];
```

Comment:
This will not work in the case of a multi-root workspace.
Errors from a secondary root will not be displayed in the tabbar.

Reply:
In the multi-root case, do you know if there is a way to get the root from the partial URI (title.caption)?

Reply:
> In the multi-root case, do you know if there is a way to get the root from the partial URI (title.caption)?

I'm not sure, worst case there may be two workspaces with similar structure.
Something like `README.md` at the root of the repo may be hard to find the appropriate root for (since there are a lot of projects with that structure).

Reply:
> >

It seems right now the `tabbar` does not have an appropriate multi-root support. The tab data defined in phosphorJS has no `Path`/`URI` field, only the `title.caption` that returns the incomplete path (e.g., root1/README.md => README.md, root2/index.ts => index.ts).

Reply:
If there is a multi-root workspace present, I'd expect that hovering over the tabbar (displaying the title) should reveal the entire path, but for the moment we only reveal the relative path. In the case of 'README.md' we only display 'README.md' for both roots.

This should be fixed, I'll file an issue for it.

Reply:
@fangnx I fixed and merged the tabbar (https://github.com/theia-ide/theia/pull/5924), please make sure to rebase the PR and confirm if it still works correctly :)